### PR TITLE
feat: use url encoding type for list calls

### PR DESCRIPTION
**Pull request by pranavosu**
#### Description of changes

- Set [encoding type](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#AmazonS3-ListObjectsV2-response-EncodingType) to `url` for list requests.
- Decode output values.
- Added unit tests

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

#### Checklist for repo maintainers

- [ ] Verify E2E tests for existing workflows are working as expected or add E2E tests for newly added workflows
- [ ] New source file paths included in this PR have been added to CODEOWNERS, if appropriate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/providers/s3/types/options.ts [hunk 1]
```diff
@@ -57,6 +57,14 @@ interface CommonOptions {
...
 	 */
 	expectedBucketOwner?: string;
+
+	/**
+	 * Specifies the encoding used for response elements that contain
+	 * special characters. By setting this to 'url', special characters
+	 * in object keys are URL-encoded in the response using the
+	 * `application/x-www-form-urlencoded` encoding.
+	 */
+	encodingType?: 'url';
```
Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/types/options.ts:67]: Do we need to expose it to all library users? Can we just set the encoding type to url for all requests?

File: packages/storage/__tests__/providers/s3/apis/internal/list.test.ts [hunk 2]
```diff
@@ -1024,4 +1024,35 @@ describe('list API', () => {
 			});
 		});
 	});
+
+	describe.only.each([
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/apis/internal/list.test.ts:1028]: `.only` to be removed

File: packages/storage/src/providers/s3/utils/client/s3data/listObjectsV2.ts [hunk 3]
```diff
@@ -155,6 +157,48 @@ const validateCorroboratingElements = (response: ListObjectsV2Output) => {
...
 };
 
+/**
+ * Decodes URL-encoded elements in the S3 `ListObjectsV2Output` response when `EncodingType` is `'url'`.
+ * Applies to values for 'Delimiter', 'Prefix', 'StartAfter' and 'Key' in the response.
+ */
+const decodeEncodedElements = (
+	listOutput: ListObjectsV2Output,
+): ListObjectsV2Output => {
+	if (listOutput.EncodingType !== 'url') {
```
Comment (ashika112) [inline packages/storage/src/providers/s3/utils/client/s3data/listObjectsV2.ts:167]: Can this be not `url` if we are the ones setting this in first place?

Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/utils/client/s3data/listObjectsV2.ts:167]: Thanks for handling it!

Comment (pranavosu) [reply packages/storage/src/providers/s3/utils/client/s3data/listObjectsV2.ts]: I think it's highly unlikely, but I'd rather be safe.

**Review by AllanZhengYP — APPROVED** [review]
Looks good to me!

**Review by HuiSF — APPROVED** [review]

**Review by ashika112 — APPROVED** [review]

**Pull request merged**